File: native/engine/android/app/src/com/cocos/game/NativeBridge.java
package com.cocos.game;

import android.content.Context;
import android.content.pm.PackageInfo;

import java.util.Map;
import java.util.Set;

public class NativeBridge {
    // 一般来说，不需要重载不同签名的情况
    private static final Map<String, String> mMethodCache = Map.of(
            "sum", "(II)I",
            "getVersionCode", "()I",
            "getVersionName", "()Ljava/lang/String;"
    );

    private static Context mContext;
    public static void init(Context context) {
        mContext = context;
    }

    public static int getVersionCode() {
        try {
            PackageInfo packageInfo = mContext.getPackageManager().getPackageInfo(mContext.getPackageName(), 0);
            return packageInfo.versionCode;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return 0;
    }

    public static String getVersionName() {
        try {
            PackageInfo packageInfo = mContext.getPackageManager().getPackageInfo(mContext.getPackageName(), 0);
            return packageInfo.versionName;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "unknown";
    }

    public static boolean hasMethod(String methodName, String signatures) {
        return (mMethodCache.containsKey(methodName) && mMethodCache.get(methodName) == signatures);
    }
}

File: assets/scripts/framework/native/NativeBridge.ts
import { native } from 'cc';
import { ANDROID } from 'cc/env';

const CLASS_NAME = "com/cocos/game/NativeBridge";

function hasMethod(methodName: string, signatures: string): boolean {
    return native.reflection.callStaticMethod(CLASS_NAME, "hasMethod", "(Ljava/lang/String;Ljava/lang/String;)Z", methodName, signatures);
}

export function native_sum(a: number, b: number): number {
    if (ANDROID) {
        const methodName = "sum", signatures = "(II)I";
        if (hasMethod(methodName, signatures)) {
            return native.reflection.callStaticMethod(CLASS_NAME, methodName, signatures, a, b);
        }
    }

    return 0;
}

export function native_getVersionName(): string {
    if (ANDROID) {
        const methodName = "getVersionName", signatures = "()Ljava/lang/String;";
        if (hasMethod(methodName, signatures)) {
            return native.reflection.callStaticMethod(CLASS_NAME, methodName, signatures);
        }
    }
    return "unknown";
}
